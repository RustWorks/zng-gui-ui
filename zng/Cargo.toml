[package]
name = "zng"
version = "0.1.0"
authors = ["Samuel Guerra <sam.rodr.g@gmail.com>", "Well <well-r@hotmail.com>"]
edition = "2021"
license = "Apache-2.0"
readme = "README.md"

[features]
default = ["ipc", "debug_default", "view_software"]

# Enables pre-build views and connecting to views running in another process.
# 
# Enabled by default.
ipc = ["zng-app/ipc", "zng-view-api/ipc", "zng-view?/ipc"]

# Enables http tasks.
http = ["zng-task/http", "zng-wgt/http", "zng-ext-image/http"]

# Enable the `dyn_*`, `inspector` features for debug builds only.
#
# Enabled by default.
debug_default = [
    "zng-app/debug_default",
    "zng-wgt-inspector/debug_default",
    "zng-wgt-scroll/debug_default",
    "zng-wgt-window/debug_default",
]

# Use more dynamic dispatch at the node level by enabling `UiNode::cfg_boxed` to box.
#
# This speeds-up compilation time at the cost of runtime.
dyn_node = ["zng-app/dyn_node"]

# Instrument each property and widget instance with "Inspector" nodes and
# extend windows to be inspected on Ctrl+Shift+I.
inspector = ["zng-app/inspector", "zng-wgt-inspector/live"]

# Use dynamic dispatch at the app-extension level.
#
# This speeds-up compilation time at the cost of runtime.
dyn_app_extension = ["zng-app/dyn_app_extension"]

# Box closures at opportune places, such as `Var::map`, reducing the number of monomorphised types.
#
# This speeds-up compilation time at the cost of runtime.
dyn_closure = [
    "zng-wgt/dyn_closure",
    "zng-wgt-scroll/dyn_closure",
    "zng-wgt-window/dyn_closure",
]

# Test utilities.
#
# USAGE:
#
# Build with this feature to use the core test helper types that show
# in the documentation.
test_util = ["zng-app/test_util", "zng-ext-window/test_util"]

# Allows multiple app instances per-process.
#
# This feature allows multiple apps, one app per thread at a time. The `LocalContext` tracks
# what app is currently running in each thread and `app_local!` statics switch to the value of each app
# depending on the current thread.
#
# Not enabled by default, but enabled by `feature="test_util"`.
multi_app = ["zng-app/multi_app"]

# Instrument every widget outer-most node to trace UI methods.
trace_widget = ["zng-app/trace_widget"]

# Instrument every property and intrinsic node to trace UI methods.
#
# Note that this can cause very large trace files and bad performance.
trace_wgt_item = ["zng-app/trace_wgt_item"]

# Spawns a thread on app creation that checks and prints `parking_lot` deadlocks.
#
# Not enabled by default.
deadlock_detection = ["zng-app/deadlock_detection"]

# Embed hyphenation dictionaries for all supported languages.
#
# If enabled some 2.8MB of data is embedded, you can provide an alternative dictionary source using the 
# `Hyphenation::dictionary_source` method.
#
# Not enabled by default.
hyphenation_embed_all = ["zng-ext-font/hyphenation_embed_all"]

# Include all Material Icons icon sets in the default app.
#
# Not enabled by default.
material_icons = [
    "material_icons_outlined",
    "material_icons_filled",
    "material_icons_rounded",
    "material_icons_sharp",
]
# Material Icons Outlined icon set.
#
# If enabled some icons of this set are used for some of the commands.
#
# Not enabled by default.
material_icons_outlined = [
    "zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/outlined",
]

# Material Icons Filled icon set.
#
# Not enabled by default.
material_icons_filled = [
    "zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/filled",
]

# Material Icons Rounded icon set.
#
# Not enabled by default.
material_icons_rounded = [
    "zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/rounded",
]

# Material Icons Sharp icon set.
#
# Not enabled by default.
material_icons_sharp = [
    "zng-wgt-material-icons",
    "zng-wgt-material-icons/embedded",
    "zng-wgt-material-icons/sharp",
]

# Enable TOML configs.
toml = ["zng-ext-config/toml"]
# Enable RON configs.
ron = ["zng-ext-config/ron"]
# Enable YAML configs.
yaml = ["zng-ext-config/yaml"]

# Include the default view-process implementation.
view = ["zng-view"]
# Enables software renderer fallback in the default view-process.
#
# If enabled and a native OpenGL 3.2 driver is not available the `swgl` software renderer is used.
#
# Feature enabled by default.
view_software = ["zng-view?/software"]

# Include the default view-process implementation as an embedded precompiled binary.
view_prebuilt = ["zng-view-prebuilt", "ipc"]

[dependencies]
# foundation
zng-clone-move = { path = "../zng-clone-move" }
zng-txt = { path = "../zng-txt" }
zng-state-map = { path = "../zng-state-map" }
zng-app-context = { path = "../zng-app-context" }
zng-layout = { path = "../zng-layout" }
zng-var = { path = "../zng-var" }
zng-task = { path = "../zng-task" }
zng-color = { path = "../zng-color" }

# view
zng-view-api = { path = "../zng-view-api", default-features = false }
zng-view = { path = "../zng-view", default-features = false, optional = true }
zng-view-prebuilt = { path = "../zng-view-prebuilt", default-features = false, optional = true }

# app
zng-app = { path = "../zng-app" }
zng-ext-fs-watcher = { path = "../zng-ext-fs-watcher" }
zng-ext-config = { path = "../zng-ext-config", default-features = false }
zng-ext-font = { path = "../zng-ext-font" }
zng-ext-l10n = { path = "../zng-ext-l10n" }
zng-ext-image = { path = "../zng-ext-image" }
zng-ext-clipboard = { path = "../zng-ext-clipboard" }
zng-ext-window = { path = "../zng-ext-window" }
zng-ext-input = { path = "../zng-ext-input" }
zng-ext-undo = { path = "../zng-ext-undo" }

# widgets
zng-wgt = { path = "../zng-wgt" }
zng-wgt-access = { path = "../zng-wgt-access" }
zng-wgt-transform = { path = "../zng-wgt-transform" }
zng-wgt-input = { path = "../zng-wgt-input" }
zng-wgt-data = { path = "../zng-wgt-data" }
zng-wgt-filter = { path = "../zng-wgt-filter" }
zng-wgt-size-offset = { path = "../zng-wgt-size-offset" }
zng-wgt-container = { path = "../zng-wgt-container" }
zng-wgt-undo = { path = "../zng-wgt-undo" }
zng-wgt-data-view = { path = "../zng-wgt-data-view" }
zng-wgt-fill = { path = "../zng-wgt-fill" }
zng-wgt-style = { path = "../zng-wgt-style" }
zng-wgt-checkerboard = { path = "../zng-wgt-checkerboard" }
zng-wgt-window = { path = "../zng-wgt-window" }
zng-wgt-layer = { path = "../zng-wgt-layer" }
zng-wgt-undo-history = { path = "../zng-wgt-undo-history" }
zng-wgt-image = { path = "../zng-wgt-image" }
zng-wgt-text = { path = "../zng-wgt-text" }
zng-wgt-text-input = { path = "../zng-wgt-text-input" }
zng-wgt-button = { path = "../zng-wgt-button" }
zng-wgt-stack = { path = "../zng-wgt-stack" }
zng-wgt-panel = { path = "../zng-wgt-panel" }
zng-wgt-grid = { path = "../zng-wgt-grid" }
zng-wgt-wrap = { path = "../zng-wgt-wrap" }
zng-wgt-rule-line = { path = "../zng-wgt-rule-line" }
zng-wgt-toggle = { path = "../zng-wgt-toggle" }
zng-wgt-menu = { path = "../zng-wgt-menu" }
zng-wgt-scroll = { path = "../zng-wgt-scroll" }
zng-wgt-ansi-text = { path = "../zng-wgt-ansi-text" }
zng-wgt-tooltip = { path = "../zng-wgt-tooltip" }
zng-wgt-markdown = { path = "../zng-wgt-markdown" }
zng-wgt-inspector = { path = "../zng-wgt-inspector" }

zng-wgt-material-icons = { path = "../zng-wgt-material-icons", default-features = false, optional = true }

[package.metadata.docs.rs]
rustdoc-args = [
    "--html-in-header",
    "target/doc/zng-material-icons-extensions.html",
]