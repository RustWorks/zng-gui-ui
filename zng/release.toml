pre-release-replacements = [
  {file="../README.md", search="zng = \\{ version = .*, ", replace="{{crate_name}} = { version = \"{{version}}\", "},
  {file="src/lib.rs", search="zng = \\{ version = .*, ", replace="{{crate_name}} = { version = \"{{version}}\", "},
  {file="src/app.rs", search="zng = \\{ version = .*, ", replace="{{crate_name}} = { version = \"{{version}}\", "},
  {file="src/icon.rs", search="zng = \\{ version = .*, ", replace="{{crate_name}} = { version = \"{{version}}\", "},
  {file="../zng-view/src/lib.rs", search="zng = .*", replace="{{crate_name}} = \"{{version}}\""},
]