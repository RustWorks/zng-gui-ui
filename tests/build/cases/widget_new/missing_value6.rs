use zng::{
    layout::margin,
    widget::{background_gradient, Wgt},
    APP,
};

fn main() {
    let _scope = APP.minimal();
    let _ = Wgt! {
        background_gradient = 0.deg(),
        margin = 0;
    };
}
