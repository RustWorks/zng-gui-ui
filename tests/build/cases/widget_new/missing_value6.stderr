error: expected `;`
 --> cases/widget_new/missing_value6.rs:8:9
  |
8 |         margin = 0;
  |         ^^^^^^

error[E0061]: this function takes 2 arguments but 1 argument was supplied
 --> cases/widget_new/missing_value6.rs:7:9
  |
7 |         background_gradient = 0.deg(),
  |         ^^^^^^^^^^^^^^^^^^^ an argument is missing
  |
note: associated function defined here
 --> $WORKSPACE/zero-ui/src/properties/visual.rs
  |
  | / #[property(fill, default(0.deg(), {
  | |     let c = colors::BLACK.transparent();
  | |     crate::core::gradient::stops![c, c]
  | | }))]
  | |____^
  = note: this error originates in the attribute macro `property` (in Nightly builds, run with -Z macro-backtrace for more info)
help: provide the argument
    |
7   |         background_gradient(=, /* value */) = 0.deg(),
    |                            ++++++++++++++++
